bgpv1:  Local internalTrafficPolicy support for ClusterIP advertisements

chaunceyjiang opened the pull request:
Fixes: #31389

```release-note
bgpv1: Add Local internalTrafficPolicy support for ClusterIP advertisements
```

harsimran-pabla commented on pkg/bgpv1/manager/reconciler/service.go:
```diff
@@ -129,11 +129,6 @@ endpointsLoop:
 			continue
 		}
 
-		// We only need Endpoints tracking for externalTrafficPolicy=Local
```
I am not sure if I understand why this is being removed ?

By removing it, all endpoints will be added in `ls (localServices)`. You might need to return localServices with internalTrafficPolicy=Local, ExternalTrafficPolicy=Local and error.

chaunceyjiang replied:
My understanding is:

localServices should store all endpoints, and it should have nothing to do with what type of ExternalTrafficPolicy or InternalTrafficPolicy the svc has.

harsimran-pabla replied:
Yes, you're right. I misunderstood. Logic to check local endpoints is at line 140.

harsimran-pabla commented on pkg/bgpv1/manager/reconciler/service.go:
```diff
@@ -324,6 +319,11 @@ func (r *ServiceReconciler) svcDesiredRoutes(newc *v2alpha1api.CiliumBGPVirtualR
 				desiredRoutes = append(desiredRoutes, netip.PrefixFrom(addr, addr.BitLen()))
 			}
 		case v2alpha1api.BGPClusterIPAddr:
+			// Ignore internalTrafficPolicy == Local && no local endpoints.
```
I think ExternalTrafficPolicy checks should be moved into LoadBalancer or ExternalIP advertisments. Like InternalTrafficPolicy check here. 

Also, consider moving each advertisement types into separate methods, this switch case is getting rather big and each having different logic.

chaunceyjiang replied:
> I think ExternalTrafficPolicy checks should be moved into LoadBalancer or ExternalIP advertisments. Like InternalTrafficPolicy check here.


Does a clusterIP service not need to check ExternalTrafficPolicy? @harsimran-pabla

harsimran-pabla replied:
I discussed this with @rastislavs, we should be clear in documentation and consistent with how `iTP` and `eTP` configurations are consumed by BGP CP. 

Ideally, these fields are independent. But since we are advertising ClusterIP to outside world it causes some confusion regarding meaning of external/internal.

This is what I propose : 
- With ClusterIP : Only `iTP` configuration is considered for advertisement, `eTP` will be ignored. 
- ExternalIP, LBIP : Only `eTP` configuration is considered for advertisement, `iTP` will be ignored.

harsimran-pabla reviewed (changes requested):
Thanks @chaunceyjiang for taking this up. I don't think change w.r.t calculation of localService endpoints is correct. Please have another look at it.

Also, please add tests with this change.

harsimran-pabla commented on pkg/bgpv1/manager/reconciler/service_test.go:
```diff
@@ -740,14 +740,15 @@ func TestServiceReconcilerWithClusterIP(t *testing.T) {
 	svc1ETPLocal := svc1.DeepCopy()
-	svc1ETPLocal.Spec.ExternalTrafficPolicy = slim_corev1.ServiceExternalTrafficPolicyLocal
+	internalTrafficPolicyLocal := slim_corev1.ServiceInternalTrafficPolicyLocal
+	svc1ETPLocal.Spec.InternalTrafficPolicy = &internalTrafficPolicyLocal
```
nit: svc1ETPLocal -> svc1ITPLocal

harsimran-pabla replied:
Same for other variables in the test case, I am assuming it means ITP(InternalTrafficPolicy)Local.

chaunceyjiang replied:
Updated.

harsimran-pabla reviewed (approved):
Looks good, thank you for this change. I would suggest updating the document as well to clarify usage of iTP vs eTP in various service announements.

qmonnet commented on Documentation/network/bgp-control-plane.rst:
```diff
@@ -305,10 +305,15 @@ defaults to the ``LoadBalancerIP`` service. You can also specify the ``.serviceA
+the BGP Control Plane will only consider the configuration of the service's ``.spec.internalTrafficPolicy`` and ignore
+the configuration of ``.spec.externalTrafficPolicy``.
+For ``ExternalIP`` and ``LoadBalancerIP``,it will only consider the configuration of
+the service's  ``.spec.externalTrafficPolicy`` and ignore the configuration of ``.spec.internalTrafficPolicy``.
```
```suggestion
It is worth noting that when you configure ``virtualRouters[*].serviceAdvertisements`` as ``ClusterIP``,
the BGP Control Plane only considers the configuration of the service's ``.spec.internalTrafficPolicy`` and ignores
the configuration of ``.spec.externalTrafficPolicy``.
For ``ExternalIP`` and ``LoadBalancerIP``, it only considers the configuration of
the service's ``.spec.externalTrafficPolicy`` and ignores the configuration of ``.spec.internalTrafficPolicy``.
```

chaunceyjiang replied:
Updated.

qmonnet commented on Documentation/network/bgp-control-plane.rst:
```diff
@@ -305,10 +305,15 @@ defaults to the ``LoadBalancerIP`` service. You can also specify the ``.serviceA
+the BGP Control Plane will only consider the configuration of the service's ``.spec.internalTrafficPolicy`` and ignore
+the configuration of ``.spec.externalTrafficPolicy``.
+For ``ExternalIP`` and ``LoadBalancerIP``,it will only consider the configuration of
+the service's  ``.spec.externalTrafficPolicy`` and ignore the configuration of ``.spec.internalTrafficPolicy``.
```
Is it a change of the actual Control Plane's behaviour, and if so, should it be mentioned in the upgrade notes?

chaunceyjiang replied:
I don't think it's necessary, this feature was just merged and there is already very clear context about it in the description of  #31389. 

This is just a supplement to the code and docs of PR #31331.  @qmonnet

qmonnet replied:
Ok thanks!

qmonnet reviewed (changes requested):
Thanks for this work!

Could you please add the context and motivation for the change to the commit log? It's currently difficult to grasp what the commit is about without opening the related PR, which is not particularly user-friendly when looking at the code locally.

I would also recommend splitting the changes into several commits. You have a doc change that is not strictly related to the changes in the code; variables renaming in the tests; code refactoring in `svcDesiredRoutes()`; and the addition of the `internalTrafficPolicy` support in this code. Each of these could be a separate commit with a relevant description, which would make it much easier to review and to understand in the future.

I also have a few style nits, please see inline below.

qmonnet reviewed (changes requested):
Thanks for addressing my feedback on the content :+1:.

I stand by my previous comment: splitting the PR into multiple commits remains optional (although recommended), but I really want to see the context and motivation for the change in the commit log before we merge this.

chaunceyjiang committed:
bgpv1: Local internalTrafficPolicy support for ClusterIP advertisements

Fixes: #31389

As the ClusterIP services can have .spec.internalTrafficPolicy set to Local, advertising a ClusterIP from a node which has no local endpoints may cause unreachability of the advertised service if the traffic is routed to that node.

So if .spec.internalTrafficPolicy of a service is set to Local, advertise its ClusterIP over BGP only if it has non-zero local endpoints.

Signed-off-by: chaunceyjiang <chaunceyjiang@gmail.com>

chaunceyjiang committed:
bgpv1: Add the documentation for using serviceAdvertisements

Ideally, the internalTrafficPolicy and externalTrafficPolicy fields are independent. However, since we advertise the ClusterIP to the external world via BGP, it has caused some confusion regarding the meanings of external and internal. Therefore, we need to provide some clarification in the documentation.

For ClusterIP: Only consider the iTP configuration for advertisement, and ignore the eTP configuration.
For ExternalIP and LBIP: Only consider the eTP configuration for advertisement, and ignore the iTP configuration.

Signed-off-by: chaunceyjiang <chaunceyjiang@gmail.com>

qmonnet reviewed (approved):
I like it much better this way, thanks a lot! :rocket:
